File: trunk/RayTracer/RayTracer/src/Sphere.cpp
#include "Sphere.h"
#include <cmath>

Sphere::Sphere(void)
{
}

Sphere::~Sphere(void)
{
}

bool Sphere::Intersection(const ray& r, intersection_data& intersect)
{
	Vector3 result;
	Vector3 L;
	float Tca, Thc, d, T;

	L = center - r.origin;
	Tca = InnerProduct(L, r.direction);
	if (Tca < 0) return false;

	d = InnerProduct(L, L) + Tca * Tca;
	if (d > radius * radius) return false;

	Thc = sqrt(radius * radius - d);
	/* assuming T won't be negative, which should never happen in cases that matter (either we're inside the sphere and so we have bigger problems, or it's behind us)*/
	T = std::min(Tca - Thc, Tca + Thc);

	intersect.point = (Vector3)r.direction * T;
	intersect.col = this->mtl_diffuse; //change to something more complex next submission
	intersect.normal = intersect.point - this->center;
	intersect.T = T;

	return true;
}

void Sphere::BoundingBox(bounding_box& b)
{
	b.p1.x = this->center.x - radius;
	b.p1.y = this->center.y - radius;
	b.p1.y = this->center.z - radius;

	b.p2.x = this->center.x + radius;
	b.p2.y = this->center.y + radius;
	b.p2.y = this->center.z + radius;
}

File: java/src/AZBIrenderer/Sphere.java

package AZBIrenderer;

import static AZBIrenderer.Vector3.*;
import static AZBIrenderer.Math3D.*;
/**
 *
 */
public class Sphere extends Surface {

    public Point3 center;
    public float radius;

    @Override
    public boolean Intersection(Ray r, IntersectionData intersect) {
        float L_P2_P3_square, L_P2_P1, d;

        d = Point2RayDist(this.center, r);
        L_P2_P3_square = this.radius * this.radius - d * d;

        /* Now, make sure there is an intersection */
        if (L_P2_P3_square < 0 || L_P2_P3_square == Float.NaN)
        {
//            intersect.point = null;
//            intersect.col = null; //change to something more complex next submission
//            intersect.normal = null;
//            intersect.T = Float.POSITIVE_INFINITY;
            return false;
        }

        L_P2_P1 = InnerProduct(sub(center, r.origin), r.direction);

	intersect.T = L_P2_P1 - (float) Math.sqrt(L_P2_P3_square);
	intersect.point = add(r.origin, mul (intersect.T, r.direction));
	intersect.normal = Normalize (sub (intersect.point, this.center));
        intersect.col = Debug.getFromNormal(this,intersect.normal); //change to something more complex next submission
/*
        Debug.print("*****************************************");
        Debug.print(this);
        Debug.print(r);
        Debug.print("Intersection at " + Debug.makeString(intersect.point));
        Debug.print("The distance is " + Debug.makeString(intersect.T));
        Debug.print("*****************************************");
*/
	return true;
    }

    @Override
    public AZBIrenderer.BoundingBox BoundingBox() {
        return new BoundingBox(
                Vector3.add (center, radius),
                Vector3.sub (center, radius));
    }

    public void fillMissing() { }
}
